File: cracking_the_coding_interview/1_5_one_edit_away/one_edit_away.py
import unittest


# Check if both strings are identical or only one edit away
# from each other.
#
# time complexity: O(M+N)
# space complexity: O(1)
def one_edit_away(first: str, second: str) -> bool:

    # If the strings have the same length, check if they are
    # one replacement away from each other.
    if len(first) == len(second):
        return one_edit_replace(first, second)
    # If they have a length difference of one, check if they are
    # one insert away from each other.
    elif len(first) - 1 == len(second):
        return one_edit_insert(second, first)
    elif len(second) - 1 == len(first):
        return one_edit_insert(first, second)
    else:
        return False

# Check if one of the string would be edited at most one
# place, both are identical.
def one_edit_replace(first: str, second: str) -> bool:
    # Just go through both strings, return `False` as
    # soon as we find a second difference.
    found_difference = False

    for c1, c2 in zip(first, second):
        if c1 != c2:
            if found_difference:
                return False
            found_difference = True;

    return True


# Check if both strings are identical, if we
# could insert at one character on one of the strings.
def one_edit_insert(smaller: str, larger: str) -> bool:
    it1 = 0
    it2 = 0
    while it1 < len(smaller) and it2 < len(larger):
        c1 = smaller[it1]
        c2 = larger[it2]
        # If we found the first difference, advance only the iterator
        # of the larger string.
        if c1 != c2:
            # If the iterators are different, this is already the second difference.
            if it1 != it2:
                return False
            it2 += 1
        else:
            it1 += 1
            it2 += 1

    return True

# testing the implementation


class OneEditAwayTest(unittest.TestCase):
    def test_one_edit_away(self):
        test_cases = [
            ("", "", True),
            ("aa", "aa", True),
            ("pale", "ple", True),
            ("pales", "pale", True),
            ("pale", "bale", True),
            ("pale", "bae", False),
            ("abc", "ab", True),
            ("ab", "abc", True),
            ("abc", "bc", True),
            ("bc", "abc", True),
            ("abc", "a", False),
            ("a", "abc", False),
        ]
        for first, second, expected in test_cases:
            self.assertEqual(one_edit_away(first, second), expected)


if __name__ == "__main__":
    unittest.main()

File: cracking_the_coding_interview/1_5_one_edit_away/one_edit_away.rs

// Check if both strings are identical or only one edit away
// from each other.
//
// time complexity: O(M+N)
// space complexity: O(1)
fn one_edit_away(first: &str, second: &str) -> bool {
    // If the strings have the same length, check if they are
    // one replacement away from each other.
    if first.len() == second.len() {
        return one_edit_replace(first, second);
    // If they have a length difference of one, check if they are
    // one insert away from each other.
    } else if first.len() as i32 - 1 == second.len() as i32 {
        return one_edit_insert(second, first);
    } else if second.len() as i32 - 1 == first.len() as i32 {
        return one_edit_insert(first, second);
    } else {
        return false;
    }
}

// Check if one of the string would be edited at most one
// place, both are identical.
fn one_edit_replace(first: &str, second: &str) -> bool {
    let mut found_difference = false;
    for (c1, c2) in first.chars().zip(second.chars()) {
        if c1 != c2 {
            if found_difference {
                return false;
            }
            found_difference = true;
        }
    }
    true
}

// Check if both strings are identical, if we
// could insert at one character on one of the strings.
fn one_edit_insert(smaller: &str, larger: &str) -> bool {
    let mut it1 = smaller.chars();
    let mut it2 = larger.chars();
    let mut found_difference = false;

    while let (Some(c1), Some(c2)) = (it1.next(), it2.next()) {
        if c1 != c2 {
            if found_difference {
                return false;
            }
            found_difference = true;
            // Move back one character in the smaller string's iterator.
            it1.next_back();
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_one_edit_away() {
        let test_cases = vec![
            ("", "", true),
            ("aa", "aa", true),
            ("pale", "ple", true),
            ("pales", "pale", true),
            ("pale", "bale", true),
            ("pale", "bae", false),
            ("abc", "ab", true),
            ("ab", "abc", true),
            ("abc", "bc", true),
            ("bc", "abc", true),
            ("abc", "a", false),
            ("a", "abc", false),
        ];

        for (first, second, expected) in test_cases {
            assert_eq!(one_edit_away(first, second), expected);
        }
    }
}
